### PR TITLE
Embeding files in the binary (only for release generation)

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Probably we can polish the build/release.mk to be sure we are generating
that properly.

I'm using rice, which will get the files from the directories (for
development) and from the binary for production.

At the same time I added some options to the cli to override the i18n,
email templates and client.

A good idea to review this PR is to skip the first commit only adding vendor files.

### CODE REVIEW

**Comment by lieut-data** [comment]
@jespino, haven't started the review yet, but do the files remain writable? The `subpath` feature works by rewriting some hard-coded paths in the `client/` folder to match the configured subpath. (It wasn't possible to fully parameterize things, e.g. CSS paths).

**Comment by jespino** [comment]
@lieut-data the files will be read only, but now we can use the embeded client or we can specify a --client-override parameter to specify a directory to use as client. My proposal would be allow to "extract" the files from the binary somehow (`mattermost generate client|templates|i18n [dirname]`) and then use your command to rewrite the urls, and finally use the `mattermost --client-override my-client-directory`. But I'm open to ideas about it.

**Comment by jespino** [comment]
I'm realizing that we need to handle better the default config file handling. We talk some time ago about generating automatically `config/default.json` from code. Probably that would help to remove a extra file to handle.

File: build/release.mk [hunk 1]
```diff
@@ -5,21 +5,43 @@ build-linux:
...
+build-osx:
 	@echo Build OSX amd64
 	env GOOS=darwin GOARCH=amd64 $(GO) install -i $(GOFLAGS) $(GO_LINKER_FLAGS) ./...
 
-build-windows: 
+build-windows:
 	@echo Build Windows amd64
 	env GOOS=windows GOARCH=amd64 $(GO) install -i $(GOFLAGS) $(GO_LINKER_FLAGS) ./...
 
+build-linux-embeded-static-files: build-client-for-embeded-static-files
```
Comment (ThiefMaster) [inline build/release.mk]: typo: s/embeded/embedded/

**Comment by jespino** [comment]
I added the commands to extract the client/templates/i18n and actually to generate a config.json based on the default values.

File: build/release.mk (hunk 1 above)
Comment (jespino) [reply build/release.mk]: Fixed! thanks! :)

**Comment by lieut-data** [comment]
@jespino, re: subpath assets, the server currently rewrites the client files in-place if the `SiteURL` setting configures a subpath. See https://github.com/mattermost/mattermost-server/blob/57bf7203f329c47220c457a891244e512ae868e6/utils/subpath.go#L26 for the relevant implementation. Is there a way for the server to layer this and host "replacement" files in memory?

Extracting the client files will definitely still be needed, e.g. for hosting via CDN, but that's a separate and potentially less frequent operation.

**Comment by jespino** [comment]
We can't do that directly in the same "rice" generated boxes, because they are read-only, but what you want is change the client files, so we can create a "Filesystem" interface implementation wrapping the client files box to pass it to the http.Handler. Wouldn't be hard.

**Comment by lieut-data** [comment]
Cool. Can you remind me again what the high level benefits of making these changes are? Is the plan to actually switch to distributing these all-in-one binaries vs. the `.tar.gz` bundles we're shipping today? Are there downsides?

**Comment by jespino** [comment]
The benefits are mutliples:
  - We can distribute it as a single binary.
  - You can run mattermost wherever you want (you don't need to be in an specific path to run the binary)
  - Probably you will be able to install it with go get
  - It allows more direct workflows like:
```shell
mattermost generate config my-config.json
mattermost --config config.json config set DatabaseSetting.DriverName="postgresl"
mattermost --config config.json config set DatabaseSetting.DriverName="postgres://...."
mattermost --config config.json
```
and you will have you mattermost up and running (the `config set` command is made up, but can be done easily)

The drawbacks are, the binary is way larger, and the compile time for the release is larger too (not too large, but larger).

**Comment by jespino** [comment]
Closed for now, we'll revisit it in the future.

**Pull request closed without merging**